### PR TITLE
feat(provider/azure): Add Azure Load Balancer as another type of load balancer than Azure Application Gateway

### PR DESCRIPTION
**Pull request by michaeljqzq**
* Add several Azure Load Balancer operation API in AzureNetworkClient
* Add AzureLoadBalancerProvider
* Add CreateAzureServerGroupWithAzureLoadBalancerAtomicOperation to deal with server group attached to an Azure Load Balancer
* Define the ARM template related to Load Balancer

### CODE REVIEW

**Comment by michaeljqzq** [comment]
This PR is linked with https://github.com/spinnaker/deck/pull/6955

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy [hunk 1]
```diff
@@ -351,6 +362,213 @@ class AzureNetworkClient extends AzureBaseClient {
...
+
+        log.info("Adding backend address pool to ${loadBalancer.name()} for server group ${serverGroupName}")
+      }
+
+      return "${loadBalancer.id()}/backendAddressPools/${serverGroupName}"
+    } else {
+      throw new RuntimeException("Load balancer ${loadBalancerName} not found in resource group ${resourceGroupName}")
+    }
+
+    null
```
Comment (duftler) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy]: Please use `return` statements for the groovy methods that aren't simple one-liners.

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy [hunk 2]
```diff
@@ -351,6 +362,213 @@ class AzureNetworkClient extends AzureBaseClient {
...
+      azure.loadBalancers().getByResourceGroup(resourceGroupName, loadBalancerName)
+    })
+
+    if (loadBalancer) {
+      def lbAP = loadBalancer.backends().get(serverGroupName)
+      if (lbAP) {
+        def chain = loadBalancer.update()
+          .withoutBackend(lbAP.name())
+
+        // TODO: debug only; remove this as part of the cleanup
```
Comment (duftler) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy]: Did you intend to remove this by this point?

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy [hunk 3]
```diff
@@ -351,6 +362,213 @@ class AzureNetworkClient extends AzureBaseClient {
...
+      return loadBalancer.inboundNatPools().get(serverGroupName).inner().id()
+
+    } else {
+      throw new RuntimeException("Load balancer ${loadBalancerName} not found in resource group ${resourceGroupName}")
+    }
+
+    null
+  }
+
+  // Find unused port range
```
Comment (duftler) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy]: Probably don't need this comment, as it's just the name of the method.

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/model/AzureLoadBalancer.groovy [hunk 4]
```diff
@@ -22,8 +22,33 @@ import com.netflix.spinnaker.clouddriver.azure.AzureCloudProvider
...
+import groovy.transform.CompileStatic
 
+@CompileStatic
 class AzureLoadBalancer implements LoadBalancer {
+  static enum AzureLoadBalancerType {
+    AZURE_LOAD_BALANCER,
+    AZURE_APPLICATION_GATEWAY
+
+    @Override
+    String toString() {
```
Comment (duftler) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/model/AzureLoadBalancer.groovy]: This seems like a fairly complicated routine for returning what are essentially constants. Why not just define an additional field on the enum and populate it on construction of each enum value?

Comment (duftler) [reply clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/model/AzureLoadBalancer.groovy]: Something like this (Java here for the example):

```
  public enum AzureLoadBalancerType {
    AZURE_LOAD_BALANCER("Azure Load Balancer"),
    AZURE_APPLICATION_GATEWAY("Azure Application Gateway");

    private String formattedType;

    private AzureLoadBalancerType(String formattedType) {
      this.formattedType = formattedType;
    }

    public String toString() {
      return formattedType;
    }
  }
```

**Comment by maggieneterval** [comment]
Hey @michaeljqzq, just wanted to check in since we'd like to cut the release branches as soon as possible -- let us know if you need any help pushing this across the finish line!

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy (hunk 3 above)
Comment (michaeljqzq) [reply clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy]: Refine the comment

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/model/AzureLoadBalancer.groovy (hunk 4 above)
Comment (michaeljqzq) [reply clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/model/AzureLoadBalancer.groovy]: That looks like more concise

**Comment by michaeljqzq** [comment]
> Hey @michaeljqzq, just wanted to check in since we'd like to cut the release branches as soon as possible -- let us know if you need any help pushing this across the finish line!

Thanks for reminding, I already updated with @duftler 's comments

**Comment by ezimanyi** [comment]
The travis build was failing because I merged the code formatter in this morning; I added a commit running the formatter on the new file here so that travis can pass.  (The formatter doesn't touch groovy files, which is why only one files was touched in the formatter commit.)

**Pull request merged**